src/librustc_typeck/check/coercion.rs

```diff
@@ -971,12 +1041,24 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             let fn_ptr = self.tcx.mk_fn_ptr(sig);
             let prev_adjustment = match prev_ty.kind {
                 ty::Closure(..) => Adjust::Pointer(PointerCast::ClosureFnPointer(a_sig.unsafety())),
-                ty::FnDef(..) => Adjust::Pointer(PointerCast::ReifyFnPointer),
+                ty::FnDef(..) => {
+                    if prev_drop_const {
+                        Adjust::Pointer(PointerCast::ReifyNotConstFnPointer)
+                    } else {
+                        Adjust::Pointer(PointerCast::ReifyFnPointer)
+                    }
+                }
```

Comment:
This match and the next would be changed to return an `Option` and an arm for `ty::FnPtr(..)` should be added here. `Some(Adjust::Pointer(PointerCast::NotConstFnPointer))` would be  returned if `prev_drop_const`, otherwise `None` would be returned. The `apply_adjustments` calls would be skipped if `None` is returned from the match.

`apply_adjustments` would need to be changed to handle the case where a `NotConstFnPointer` adjustment is applied when there is already a `ReifyFnPointer` adjustment (by merging them to a `NotConstFnPointer` adjustment).